controller/src/main/java/io/pravega/controller/server/retention/StreamCutBucketService.java

```diff
@@ -92,10 +93,12 @@ protected void doStart() {
                 final StreamImpl stream;
                 switch (notification.getType()) {
                     case StreamAdded:
+                        log.info("New stream {}/{} added to bucket {} ", notification.getScope(), notification.getStream(), bucketId);
                         stream = new StreamImpl(notification.getScope(), notification.getStream());
-                        retentionFutureMap.putIfAbsent(stream, getStreamRetentionFuture(stream));
+                        retentionFutureMap.computeIfAbsent(stream, x -> getStreamRetentionFuture(stream));
```

Comment:
I'm not sure how this is changing anything. In the way this is done, I can't tell the difference between using `putIfAbsent` and `computeIfAbsent` (I understand the semantic difference between the two calls).

Reply:
Lets say i have a map `Map<K, CompletableFuture<V>> map`.
And i have  a method called ```CompletableFuture<V> getFutureValue()```
Calling `map.putIfAbsent(key, getFutureValue())` `n` times will only put one entry in the map but create `n` futures. This is because we pass a value to this method and the value is computed before calling the method.  

however doing `map.computeIfAbsent(key, x -> getFutureValue())` any number of times will not perform the computation more than once so only one future is created. This is because we pass a lambda here which is executed only if computation should be done. 

Note: creating future implies creating some asynchronous processing work.